### PR TITLE
Added Order logic

### PR DESCRIPTION
**Pull request by Nazar090**

### CODE REVIEW

File: src/main/java/com/example/onlinebookstore/dto/order/OrderRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package com.example.onlinebookstore.dto.order;
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class OrderRequestDto {
+    private String shippingAddress;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/dto/order/OrderRequestDto.java:13]: add validation

File: src/main/java/com/example/onlinebookstore/dto/order/OrderStatusDto.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.example.onlinebookstore.dto.order;
+
+import com.example.onlinebookstore.model.Order;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class OrderStatusDto {
+    private Order.Status status;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/dto/order/OrderStatusDto.java:15]: add validation

File: src/main/java/com/example/onlinebookstore/model/Order.java [hunk 3]
```diff
@@ -1,4 +1,49 @@
...
+@Entity
+@Table(name = "orders")
 public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(optional = false, fetch = FetchType.LAZY)
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+    @Column(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/Order.java:34]:
```suggestion
    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
```

File: src/main/java/com/example/onlinebookstore/model/Order.java [hunk 4]
```diff
@@ -1,4 +1,49 @@
...
+    private Status status = Status.PROCESSING;
+    @Column(nullable = false)
+    private BigDecimal total;
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/Order.java]:
```suggestion
    private LocalDateTime orderDate = LocalDateTime.now();
```

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    private final OrderItemMapper orderItemMapper;
+    private final CartItemRepository cartItemRepository;
+
+    @Transactional
+    @Override
+    public OrderDto createOrder(Long userId, OrderRequestDto requestDto) {
+        Order order = orderMapper.toEntity(requestDto);
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId);
+
+        order.setUser(userRepository.findById(userId)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java]: Could it be that user not found here?

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    @Override
+    public List<OrderItemDto> getOrderItemsByOrderId(Long orderId) {
+        Order order = orderRepository.findById(orderId)
+                .orElseThrow(() -> new EntityNotFoundException("Order not found with id: "
+                        + orderId));
+        return order.getOrderItems().stream()
+                .map(orderItemMapper::toDto)
+                .collect(Collectors.toList());
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java]:
```suggestion
                .toList();
```

MapStruct works well with list actually

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
let's fix some comments

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    private final OrderItemMapper orderItemMapper;
+    private final CartItemRepository cartItemRepository;
+
+    @Transactional
+    @Override
+    public OrderDto createOrder(Long userId, OrderRequestDto requestDto) {
+        Order order = orderMapper.toEntity(requestDto);
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId);
+
+        order.setUser(userRepository.findById(userId).orElseThrow());
```
Comment (ahoienko) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java]:
```suggestion
        order.setUser(userRepository.findById(userId).orElseThrow(() -> new EntityNotFoundException("Can`t find user with id: " + id)));
```

Comment (ahoienko) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java]: I believe user could be taken from shopping cart

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final OrderMapper orderMapper;
+    private final OrderItemMapper orderItemMapper;
+    private final CartItemRepository cartItemRepository;
+
+    @Transactional
+    @Override
+    public OrderDto createOrder(Long userId, OrderRequestDto requestDto) {
+        Order order = orderMapper.toEntity(requestDto);
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId);
```
Comment (ahoienko) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java]: handle case when shopping cart is missing

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        order.setTotal(getTotal(shoppingCart));
+        order.setOrderDate(LocalDateTime.now());
+        Set<OrderItem> orderItems = shoppingCart.getCartItems().stream()
+                .map(cartItem -> {
+                    OrderItem orderItem = new OrderItem();
+                    orderItem.setOrder(order);
+                    orderItem.setBook(cartItem.getBook());
+                    orderItem.setQuantity(cartItem.getQuantity());
+                    orderItem.setPrice(cartItem.getBook().getPrice());
+                    return orderItem;
```
Comment (ahoienko) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java]: I suggest to move this logic to separate method or use mapper

**Review by ahoienko — CHANGES_REQUESTED** [review]
Nice, left minor comments

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    private final OrderMapper orderMapper;
+    private final OrderItemMapper orderItemMapper;
+    private final CartItemRepository cartItemRepository;
+
+    @Transactional
+    @Override
+    public OrderDto createOrder(Long userId, OrderRequestDto requestDto) {
+        Order order = orderMapper.toEntity(requestDto);
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId);
+        if (shoppingCart == null) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java]: for such options it's better to use Optional 

but could it be the case, when shop cart null for user? If you create shop cart during user registration, it's OneToOne mapping, I don't think we should check that cart is null
But this is not mistake

File: src/main/java/com/example/onlinebookstore/validation/StatusValidator.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+import com.example.onlinebookstore.model.Order;
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+
+public class StatusValidator implements ConstraintValidator<ValidStatus, Order.Status> {
+    @Override
+    public boolean isValid(Order.Status status, ConstraintValidatorContext context) {
+        try {
+            Order.Status.valueOf(String.valueOf(status));
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/validation/StatusValidator.java:11]: what the purpose of this validator?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/onlinebookstore/validation/StatusValidator.java (hunk 11 above)
Comment (Nazar090) [reply src/main/java/com/example/onlinebookstore/validation/StatusValidator.java:11]: check if the status of the request is correct

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**